File: sdk/js-web-holons/src/discovery_types.mjs
/**
 * Shared discovery constants and JSDoc types for js-web-holons.
 */

export const LOCAL = 0;
export const PROXY = 1;
export const DELEGATED = 2;

export const SIBLINGS = 0x01;
export const CWD = 0x02;
export const SOURCE = 0x04;
export const BUILT = 0x08;
export const INSTALLED = 0x10;
export const CACHED = 0x20;
export const ALL = 0x3F;

export const NO_LIMIT = 0;
export const NO_TIMEOUT = 0;

/**
 * @typedef {Object} IdentityInfo
 * @property {string} given_name
 * @property {string} family_name
 * @property {string} [motto]
 * @property {string[]} [aliases]
 */

/**
 * @typedef {Object} HolonInfo
 * @property {string} slug
 * @property {string} uuid
 * @property {IdentityInfo} identity
 * @property {string} lang
 * @property {string} runner
 * @property {string} status
 * @property {string} kind
 * @property {string} transport
 * @property {string} entrypoint
 * @property {string[]} architectures
 * @property {boolean} has_dist
 * @property {boolean} has_source
 */

/**
 * @typedef {Object} HolonRef
 * @property {string} url
 * @property {HolonInfo|null} info
 * @property {string|null} error
 */

/**
 * @typedef {Object} DiscoverResult
 * @property {HolonRef[]} found
 * @property {string|null} error
 */

/**
 * @typedef {Object} ResolveResult
 * @property {HolonRef|null} ref
 * @property {string|null} error
 */

/**
 * @typedef {Object} ConnectResult
 * @property {object|null} channel
 * @property {string} uid
 * @property {HolonRef|null} origin
 * @property {string|null} error
 */

File: sdk/csharp-holons/Holons/DiscoveryTypes.cs
namespace Holons;

public static partial class Discovery
{
    public const int LOCAL = 0;
    public const int PROXY = 1;
    public const int DELEGATED = 2;

    public const int SIBLINGS = 0x01;
    public const int CWD = 0x02;
    public const int SOURCE = 0x04;
    public const int BUILT = 0x08;
    public const int INSTALLED = 0x10;
    public const int CACHED = 0x20;
    public const int ALL = 0x3F;

    public const int NO_LIMIT = 0;
    public const int NO_TIMEOUT = 0;
}

public sealed record IdentityInfo(
    string GivenName,
    string FamilyName,
    string Motto,
    IReadOnlyList<string> Aliases);

public sealed record HolonInfo(
    string Slug,
    string Uuid,
    IdentityInfo Identity,
    string Lang,
    string Runner,
    string Status,
    string Kind,
    string Transport,
    string Entrypoint,
    IReadOnlyList<string> Architectures,
    bool HasDist,
    bool HasSource);

public sealed record HolonRef(string Url, HolonInfo? Info, string? Error);

public sealed record DiscoverResult(IReadOnlyList<HolonRef> Found, string? Error);

public sealed record ResolveResult(HolonRef? Ref, string? Error);

public sealed record ConnectResult(object? Channel, string UID, HolonRef? Origin, string? Error);
